crates/cargo-gpu/src/spirv_cli.rs

```diff
@@ -1,64 +1,70 @@
-//! Install the relevant Rust compiler and toolchain components required by the version of
-//! `rust-gpu` defined in the shader.
-use crate::cache_dir;
+//! Query the shader crate to find what version of `rust-gpu` it depends on.
+//! Then ensure that the relevant Rust toolchain and components are installed.
 
-/// The canonical pairs of `rust-gpu` version to `rust-gpu` Rust toolchain channel.
-///
-/// TODO: We may be able to automate this by checking out the locked version of `spirv-std` in the
-/// shader's crate.
-const SPIRV_STD_TOOLCHAIN_PAIRS: &[(&str, &str)] = &[("0.10", "nightly-2024-04-24")];
+use crate::spirv_source::SpirvSource;
 
 /// Cargo dependency for `spirv-builder` and the rust toolchain channel.
 #[derive(Debug, Clone)]
-pub struct Spirv {
-    /// The version of `rust-gpu`, eg `"0.10"` or `"{ git = "https://github.com/Rust-GPU/rust-gpu.git" }`
-    pub dep: String,
-    //// The toolchain that `rust-gpu` uses, eg "nightly-2024-04-24"
+pub struct SpirvCLI {
+    #[expect(
+        clippy::doc_markdown,
+        reason = "The URL should appear literally like this. But Clippy wants it to be a in markdown clickable link"
+    )]
+    /// The source and version of `rust-gpu`.
+    /// Eg:
+    ///   * From crates.io with version "0.10.0"
+    ///   * From Git with:
+    ///     - a repo of "https://github.com/Rust-GPU/rust-gpu.git"
+    ///     - a revision of "abc213"
+    pub source: SpirvSource,
+    /// The toolchain channel that `rust-gpu` uses, eg "nightly-2024-04-24"
     pub channel: String,
+    /// The date of the pinned version of `rust-gpu`
```

Comment:
Is this the date of `rust-gpu` or the date of the nightly toolchain we need to build `spirv-builder-cli` (and the shader crat) with?

Reply:
It's the date of `rust-gpu`, using that gives us precise control over compatibility.